[FX qconfig] add weighted_op_qint8_dtype_config for int8 TRT and import linear config to get_tensorrt_backend_config_dict

brad-mengchi opened the pull request:
Summary:
- Add weighted_op_qint8_dtype_config
- import import linear config to get_tensorrt_backend_config_dict()
- Add qconfig for _get_linear_configs()

Test Plan: CI

Differential Revision: D36152319

jerryzh168 commented on torch/ao/quantization/backend_config/native.py:
```diff
@@ -53,6 +53,16 @@
 
 # weighted op int8 dtype config
 # this is config for ops that has quantized weights, like linear, conv
+weighted_op_qint8_dtype_config = {
```
we can remove this I think

jerryzh168 commented on torch/ao/quantization/backend_config/native.py:
```diff
@@ -692,6 +695,14 @@ def _get_embedding_op_configs():
 def get_native_backend_config_dict():
     """ Get backend_config_dict for PyTorch Native backend (fbgemm/qnnpack). """
+    linear_op_dtype_configs = [
+        weighted_op_qint8_dtype_config,
```
remove this one?

jerryzh168 commented on torch/ao/quantization/backend_config/native.py:
```diff
@@ -692,6 +695,14 @@ def _get_embedding_op_configs():
 
 def get_native_backend_config_dict():
     """ Get backend_config_dict for PyTorch Native backend (fbgemm/qnnpack). """
+    linear_op_dtype_configs = [
```
nit: linear_op_dtype_configs -> linear_dtype_configs

jerryzh168 commented on torch/ao/quantization/backend_config/tensorrt.py:
```diff
@@ -32,22 +33,48 @@ def get_tensorrt_backend_config_dict():
         # optional, output activation dtype
         "output_dtype": torch.qint8,
     }
+    weighted_op_int8_dtype_config = {
+        # optional, input activation dtype
+        "input_dtype": torch.quint8,
+        # optional, weight dtype
+        "weight_dtype": torch.qint8,
+        # optional, bias dtype
+        "bias_dtype": torch.float,
+        # optional, output activation dtype
+        "output_dtype": torch.quint8
+    }
+    default_op_fp16_dtype_config = {
+        # optional, input activation dtype
+        "input_dtype": torch.float16,
+        # optional, weight dtype
+        "weight_dtype": torch.float16,
+        # optional, bias dtype
+        "bias_dtype": torch.float16,
+        # optional, output activation dtype
+        "output_dtype": torch.float16,
+    }
+    default_dynamic_int8_dtype_config = {
+        "input_dtype": torch.quint8,
+        "weight_dtype": torch.qint8,
+        "bias_dtype": torch.float,
+        "output_dtype": torch.float,
+        # currently the dtype check is not yet enabled, so we provided the dtype_configs but
+        # it is not really used yet,
+        # we will enable it a bit later after we moved everything to backend_config_dict
+        "is_dynamic": True,
+    }
+    default_dynamic_float16_dtype_config = {
+        "input_dtype": torch.float16,
+        "weight_dtype": torch.float16,
+        "bias_dtype": torch.float,
+        "output_dtype": torch.float,
```
we can remove all configs here I think

jerryzh168 commented on torch/ao/quantization/backend_config/tensorrt.py:
```diff
@@ -193,6 +220,14 @@ def get_tensorrt_backend_config_dict():
         ]
     }
+    linear_op_dtype_configs = [
+        weighted_op_qint8_dtype_config,
+        weighted_op_int8_dtype_config,
+        default_dynamic_int8_dtype_config,
+        default_dynamic_float16_dtype_config,
+        # TODO: maybe remove this since fbgemm/qnnpack doesn't have kernels for it
+        default_op_fp16_dtype_config,
```
we can remove everything here and just keep `weighted_op_qint8_dtype_config` since that's the only thing supported by tensorrt

jerryzh168 commented on torch/ao/quantization/backend_config/tensorrt.py:
```diff
@@ -193,6 +220,14 @@ def get_tensorrt_backend_config_dict():
             non_weighted_op_qint8_dtype_config,
         ]
     }
+    linear_op_dtype_configs = [
```
nit: linear_op -> linear

jerryzh168 commented on torch/ao/quantization/backend_config/tensorrt.py:
```diff
@@ -34,20 +35,6 @@ def get_tensorrt_backend_config_dict():
-        "reference_quantized_module_for_root": torch.nn.quantized._reference.Linear,
-        "qat_module": nnqat.Linear,
-    }
     linear_qat_config = {
```
I think we can get rid of all linear related configs in this file

jerryzh168 commented on torch/ao/quantization/backend_config/tensorrt.py:
```diff
@@ -203,7 +193,6 @@ def get_tensorrt_backend_config_dict():
         "name": "tensorrt",
         "configs": [
-            linear_module_config,
             linear_qat_config,
```
we can get rid of all linear configs here I think

jerryzh168 reviewed (approved):
looks good, thanks! had one additional comment to remove linear related configs in tensorrt.py

Commit:
[FX qconfig] add weighted_op_qint8_dtype_config for int8 TRT and import linear config to get_tensorrt_backend_config_dict (#76877)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/76877

- Add weighted_op_qint8_dtype_config
- import linear config to get_tensorrt_backend_config_dict()
- Add qconfig for _get_linear_configs()

Test Plan: CI

Reviewed By: jerryzh168

Differential Revision: D36152319

fbshipit-source-id: 6a14c4f37e830a01333ff1e1706d159d038e14eb
